SHARD-2309 : adds delay to SyncV2 operations

abdulazeem-tk4vr opened the pull request:
### **User description**
Introduces configurable delay to SyncV2 operations to debug potential issues.

The delay can be enabled/disabled and its duration is configurable through core flags.


___

### **PR Type**
enhancement


___

### **Description**
- Add configurable delay to SyncV2 protocol operations

- Introduce `delaySyncV2` and `enableSyncV2Delay` flags in core flags

- Apply delay in key SyncV2 functions for debugging

- Update type definitions and defaults for new flags


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Add conditional delay to SyncV2 operations using core flags</code></dd></summary>
<hr>

src/p2p/SyncV2/index.ts

<li>Import <code>CoreFlags</code> for configuration access<br> <li> Insert conditional sleep in key SyncV2 functions based on new flags<br> <li> Enable debugging of SyncV2 with configurable delay


</details>


  </td>
  <td><a href="https://github.com/shardeum/core/pull/477/files#diff-2674a10db5cab73526e3e2a7d8be5db01defb64bd55bd065f5920093ecbc649a">+4/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>coreFlags.ts</strong><dd><code>Add delay configuration flags to core flags definition</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/shardus/coreFlags.ts

<li>Extend <code>CoreFlags</code> interface with <code>delaySyncV2</code> and <code>enableSyncV2Delay</code><br> <li> Set default values for new delay-related flags<br> <li> Document new flags for clarity


</details>


  </td>
  <td><a href="https://github.com/shardeum/core/pull/477/files#diff-1bd141b288dfd485e4ed849464336021286cf590c1ba2e114a471efeb2a9ae49">+10/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about PR-Agent usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

abdulazeem-tk4vr committed:
feat: adds delay to SyncV2 operations

Introduces configurable delay to SyncV2 operations to debug potential issues.

The delay can be enabled/disabled and its duration is configurable through core flags.

aniketdivekar commented on src/p2p/SyncV2/index.ts:
```diff
@@ -70,7 +71,7 @@ export function syncV2(activeNodes: P2P.SyncTypes.ActiveNode[], shardus: Shardus
               )
             }
-
+            if (CoreFlags.enableSyncV2Delay) sleep(CoreFlags.delaySyncV2)
```
do we need `await` here?

abdulazeem-tk4vr committed:
feat: applies delay to data fetching in syncV2

Ensures the `enableSyncV2Delay` flag affects the data fetching process within the syncV2 function.
This change wraps data retrieval within `sleepAsync` calls when the delay is enabled, applying the delay to data retrieval from different node list and record types.

abdulazeem-tk4vr committed:
feat: introduces granular syncV2 delays

Refactors the syncV2 delay mechanism to allow for different delays
based on the type of data being synchronized (validator list, archiver
list, standby node list, tx list, latest cycle record).

This allows for more fine-grained control over sync performance and
network load, enabling the tuning of delays for each data type
independently.

jintukumardas commented on src/shardus/coreFlags.ts:
```diff
@@ -40,4 +45,11 @@ export const CoreFlags: CoreFlags = {
+  archiverSyncDelay: 120000, // CUMULATIVE DELAY is 10 minutes (2+2+2+2+2)
+  standbyNodeSyncDelay: 120000,
+  txSyncDelay: 120000,
+  latestCycleSyncDelay: 120000,
```
We might be able to group these delays under a single entity, such as SyncV2Delay or something similar.
